_posts/2024-03-22-kraft-migration.md

```diff
@@ -0,0 +1,388 @@
+---
+layout: post
+title:  "From ZooKeeper to KRaft: let's automate the migration with Strimzi"
+date: 2024-03-22
+author: paolo_patierno
+---
+
+The Apache Kafka project has been using [Apache ZooKeeper](https://zookeeper.apache.org/) since its inception to store metadata.
+Registered brokers, the current controller and topics' configuration are just a few of the data that ZooKeeper stores for supporting a running Kafka cluster.
+Using a dedicated and centralized system for maintaining configuration data sounded the right approach at that time and ZooKeeper has been doing its work for long.
+But taking care of an additional component alongside your Kafka cluster is not that simple.
+Also, ZooKeeper has started to be a limitation for better scaling in terms of topics' partitions.
+For these and more other reasons, the Apache Kafka community opened the "famous" [KIP-500](https://cwiki.apache.org/confluence/display/KAFKA/KIP-500%3A+Replace+ZooKeeper+with+a+Self-Managed+Metadata+Quorum) in late 2019 and started to work on the ZooKeeper removal.
+
+<!--more-->
+
+### What is "wrong" with ZooKeeper?
+
+As already mentioned, Kafka has been using ZooKeeper as its metadata management system for several functions:
+
+* **Cluster membership**: each broker, joining the Kafka cluster, registers itself with a temporary znode on ZooKeeper;
+* **Controller election**: when a broker starts, it tries to take the "controller" role by creating the ephemeral `/controller` znode on ZooKeeper. If it already exists, it states what is the broker in the cluster having such a role and what is its epoch;
+* **Topics configuration**: all the topics' configuration parameters are stored in ZooKeeper together with information like the number of partitions and on which brokers the replicas are hosted;
+* **Access Control Lists (ACLs)**: when a client connects to the cluster, it could be authenticated and authorized to read or write on several topics based on the ACLs stored in ZooKeeper;
```

Comment:
I was providing a little bit of context around authN and authZ. Maybe I can use your sentence instead of `based on the ACLs stored in ZooKeeper;` but not replacing the entire sentence?

Reply:
I tweaked it a little bit.